### PR TITLE
WIP: Dataflow based Preprocessing source

### PR DESCRIPTION
**Pull request by sakshamarora1**
#584

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -78,6 +78,7 @@
             "idx3 = dffml.source.idx3:IDX3Source",
             "db = dffml.source.db:DbSource",
             "ini = dffml.source.ini:INISource",
+            "preprocess = dffml.source.df:DataFlowSource",
```
Comment (pdxjohnny) [inline setup.py:81]:
```suggestion
            "df = dffml.source.df:DataFlowSource",
```

File: dffml/source/df.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    async def __aenter__(self) -> "DataFlowSourceContext":
+        self.sctx = await self.parent.source().__aenter__()
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        await self.sctx.__aexit__(exc_type, exc_value, traceback)
+
+
+@entrypoint("preprocess")
```
Comment (pdxjohnny) [inline dffml/source/df.py:49]:
```suggestion
@entrypoint("df")
```

File: tests/source/test_dfsource.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    DefFeature("Expertise", int, 1),
+    DefFeature("Trust", float, 1),
+    DefFeature("Salary", int, 1),
+)
+TEST_DATAFLOW = DataFlow(
+    operations={"edit_feature", edit_feature},
+    configs={"edit_feature": EditFeatureConfig},
+)
+# TODO Use tempfile
+TEST_SOURCE = CSVSource(CSVSourceConfig(filename="../../test.csv"))
```
Comment (pdxjohnny) [inline tests/source/test_dfsource.py:44]: Probably want to use MemorySource for testing purposes

File: tests/source/test_dfsource.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+from dffml.operation.output import GetSingle, remap
+from dffml.source.csv import CSVSource, CSVSourceConfig
+
+
+class EditFeatureConfig:
+    feature: int
+
+
+# TODO Create an operation which edits one of the features' values
+op(
```
Comment (pdxjohnny) [inline tests/source/test_dfsource.py:22]:
```suggestion
@op(
```

File: tests/source/test_dfsource.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    inputs={"feature": Definition(name="edit_feature", primitive="int")},
+    outputs={"newfeature": Definition(name="Salary", primitive="int")},
+    config_cls=EditFeatureConfig,
+)
+
```
Comment (pdxjohnny) [inline tests/source/test_dfsource.py:27]:
```suggestion
```

File: tests/source/test_dfsource.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
```
Comment (pdxjohnny) [inline tests/source/test_dfsource.py:28]:
```suggestion
```

**Comment by pdxjohnny** [comment]
@sakshamarora1 What state is this in?

**Comment by sakshamarora1** [comment]
@pdxjohnny Still in progress.
Can you help me with how should the `edit_feature` and `AssociateDefinition` operations work together? I've made `AssociateDefinition` that maps the feature name with the corresponding record value. `edit_feature` takes a value as an input and returns value*10 . How should I go about it?

**Comment by pdxjohnny** [comment]
@sakshamarora1 try adding a testcase for just the AssociateDefinition output operation.

The body of the `test_...` will look something like:

```python
feed_def = Definition(name="feed", primitive="string")
dead_def = Definition(name="dead", primitive="string")
output = Definition(name="output", primitive="string")

feed_input = Input(value="my favorite value", definition=feed_def)
face_input = Input(value="face", definition=OUTPUT, parents=[feed_input])

dead_input = Input(value="my second favorite value", definition=dead_def)
beef_input = Input(value="beef", definition=OUTPUT, parents=[dead_input])

async for ctx, results in run(DataFlow.auto(AssociateDefinition), [
    feed_input, face_input, dead_input, beef_input,
]):
    self.assertEqual(results, {
        "feed": "face",
        "dead": "beef",
    })
```

Also, you still need to check the parents in the associate operation:

```python
                parents = item.get_parents()
                for parent in parents:
                    if key == parent.definition:
                        want[key.name] = item.value # Change is here
                        break
            return want
```

If we don't check them, we will just associate all matching definitions, rather than those that are associated with a particular definition as their parent.

File: tests/source/test_dfsource.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    DefFeature("Years", int, 1),
+    DefFeature("Expertise", int, 1),
+    DefFeature("Trust", float, 1),
+    DefFeature("Salary", int, 1),
+)
+
+TEST_DATAFLOW1 = DataFlow.auto(AssociateDefinition, edit_feature)
+TEST_DATAFLOW1.seed = [
+    Input(
+        value=[feature.NAME, ""],
```
Comment (pdxjohnny) [inline tests/source/test_dfsource.py:58]:
```suggestion
        value=[feature.NAME, edit_feature.op.outputs["output"].name],
```

**Pull request merged**